File: tricky/QueensAttack2.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

public class Solution {

    // Complete the queensAttack function below.
    static int queensAttack(int n, int k, int rq, int cq, int[][] obstacles) {
        int d11, d12, d21, d22, r1, r2, c1, c2; 
        int x = rq;
        int y = cq;
        int obstPosx[] = new int[k];
        int obstPosy[] = new int[k];
      
        // Initialise the distance to end of the board. 
        d11 = Math.min( x-1, y-1 ); 
        d12 = Math.min( n-x, n-y ); 
        d21 = Math.min( n-x, y-1 ); 
        d22 = Math.min( x-1, n-y ); 
      
        r1 = y-1; 
        r2 = n-y; 
        c1 = x-1; 
        c2 = n-x; 
      
        // For each obstacle find the minimum distance. 
        // If obstacle is present in any direction, 
        // distance will be updated. 
        for (int i = 0; i < k; i++) 
        { 
            obstPosx[i] = obstacles [i][0];
            obstPosy[i] = obstacles [i][1];
            
            if ( x > obstPosx[i] && y > obstPosy[i] && 
                    x-obstPosx[i] == y-obstPosy[i] ) 
                d11 = Math.min(d11, x-obstPosx[i]-1); 
      
            if ( obstPosx[i] > x && obstPosy[i] > y && 
                    obstPosx[i]-x == obstPosy[i]-y ) 
                d12 = Math.min( d12, obstPosx[i]-x-1); 
      
            if ( obstPosx[i] > x && y > obstPosy[i] && 
                    obstPosx[i]-x == y-obstPosy[i] ) 
                d21 = Math.min(d21, obstPosx[i]-x-1); 
      
            if ( x > obstPosx[i] && obstPosy[i] > y && 
                        x-obstPosx[i] == obstPosy[i]-y ) 
                d22 = Math.min(d22, x-obstPosx[i]-1); 
      
            if ( x == obstPosx[i] && obstPosy[i] < y ) 
                r1 = Math.min(r1, y-obstPosy[i]-1); 
      
            if ( x == obstPosx[i] && obstPosy[i] > y ) 
                r2 = Math.min(r2, obstPosy[i]-y-1); 
      
            if ( y == obstPosy[i] && obstPosx[i] < x ) 
                c1 = Math.min(c1, x-obstPosx[i]-1); 
      
            if ( y == obstPosy[i] && obstPosx[i] > x ) 
                c2 = Math.min(c2, obstPosx[i]-x-1); 
        } 
      
        return d11 + d12 + d21 + d22 + r1 + r2 + c1 + c2; 
    


    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        String[] nk = scanner.nextLine().split(" ");

        int n = Integer.parseInt(nk[0]);

        int k = Integer.parseInt(nk[1]);

        String[] r_qC_q = scanner.nextLine().split(" ");

        int r_q = Integer.parseInt(r_qC_q[0]);

        int c_q = Integer.parseInt(r_qC_q[1]);

        int[][] obstacles = new int[k][2];

        for (int i = 0; i < k; i++) {
            String[] obstaclesRowItems = scanner.nextLine().split(" ");
            scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

            for (int j = 0; j < 2; j++) {
                int obstaclesItem = Integer.parseInt(obstaclesRowItems[j]);
                obstacles[i][j] = obstaclesItem;
            }
        }

        int result = queensAttack(n, k, r_q, c_q, obstacles);

        bufferedWriter.write(String.valueOf(result));
        bufferedWriter.newLine();

        bufferedWriter.close();

        scanner.close();
    }
}

File: needsOptimising/queensAttack2.py
#!/bin/python3

import math
import os
import random
import re
import sys

# Complete the queensAttack function below.
def queensAttack(n, k, rq, cq, obstacles):
    count = 0
    board = list()
    board = [[0 for x in range(n)] for y in range(n)]
    for i in range(k):
        board[obstacles[i][0]-1][obstacles[i][1]-1] = 1
    print(board)
    f1 = 0
    f2 = 0
    f3 = 0
    f4 = 0
    f5 = 0
    f6 = 0
    f7 = 0
    f8 = 0
    rq -= 1
    cq -= 1
    if cq ==0 or cq == n or rq ==0 or rq == n:
        k = n//2 +1
    else: k = n+2

    for i in range(1,k):
        if rq - i >= 0 and f1 == 0:
            if board[rq - i][cq] == 0:
                count += 1
            else:f1 = 1
        if rq + i < n and f2 == 0:
            if board[rq + i][cq] == 0:
                count += 1
            else:f2 = 1
        if cq - i >= 0 and f3 == 0:
            if board[rq][cq - i] == 0:
                count += 1
            else:f3 = 1
        if cq + i < n and f4 == 0:
            if board[rq][cq + i] == 0:
                count += 1
            else:f4 = 1
        if rq - i >= 0 and cq - i >= 0 and f5 ==0:
            if board[rq-i][cq-i] == 0:
                count += 1
            else:f5 = 1
        if rq + i < n and cq + i < n and f6 ==0:
            if board[rq+i][cq+i] == 0:
                count += 1
            else:f6 = 1
        if rq - i >= 0 and cq + i < n and f7 ==0:
            if board[rq-i][cq-i] == 0:
                count += 1
            else:f7 = 1
        if rq + i < n and cq - i >=0 and f8 ==0:
            if board[rq+i][cq+i] == 0:
                count += 1
            else:f8 = 1
                

    return count
 

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    nk = input().split()

    n = int(nk[0])

    k = int(nk[1])

    r_qC_q = input().split()

    r_q = int(r_qC_q[0])

    c_q = int(r_qC_q[1])

    obstacles = []

    for _ in range(k):
        obstacles.append(list(map(int, input().rstrip().split())))

    result = queensAttack(n, k, r_q, c_q, obstacles)

    fptr.write(str(result) + '\n')

    fptr.close()
